File: scripts/s07_oracle/graph_stream.go
package main

import (
	"encoding/hex"
	"encoding/json"
	"github.com/microsoft/TypeScript/tsc/internal/ast"
)

func dumpGraph(s *session, stage string, file *ast.SourceFile) {
	for _, record := range collectGraph(file) {
		observeGraph(s, stage, record.kind, record.value)
	}
}
func observeGraph(s *session, stage, kind string, value any) {
	const chunk = 128 * 1024
	raw, err := json.Marshal(value)
	if err != nil {
		panic(err)
	}
	if len(raw) <= chunk {
		s.observe(stage, kind, value)
		return
	}
	parts := (len(raw) + chunk - 1) / chunk
	for part := 0; part < parts; part++ {
		end := min((part+1)*chunk, len(raw))
		s.observe(stage, "fragment", map[string]any{"record_kind": kind, "part": part, "parts": parts, "payload_hex": hex.EncodeToString(raw[part*chunk : end])})
	}
}

File: crates/ts_binder/examples/support/graph_stream.rs
use super::{graph, protocol::Session};
use serde_json::{json, Value};
use ts_ast::{AstView, BindResult, NodeId};

pub fn dump(
    s: &Session,
    stage: &str,
    view: AstView<'_>,
    source: NodeId,
    result: Option<&BindResult>,
) {
    for (kind, value) in graph::collect(view, source, result) {
        observe(s, stage, kind, value);
    }
}
fn observe(s: &Session, stage: &str, kind: &str, value: Value) {
    const CHUNK: usize = 128 * 1024;
    let raw = match serde_json::to_vec(&value) {
        Ok(raw) => raw,
        Err(error) => {
            s.failure(format!("S07 graph serialization: {error}"));
            return;
        }
    };
    if raw.len() <= CHUNK {
        s.observe(stage, kind, value);
        return;
    }
    let parts = raw.len().div_ceil(CHUNK);
    for (part, bytes) in raw.chunks(CHUNK).enumerate() {
        s.observe(
            stage,
            "fragment",
            json!({"record_kind":kind,"part":part,"parts":parts,"payload_hex":graph::hex(bytes)}),
        );
    }
}
